### PR TITLE
PWGHF: add ML for D* and MC info in charm polarisation task 

### PR DESCRIPTION
**Pull request by fgrosa**
This PR adds:
- the possibility to apply ML selections for D* and store the ML outputs in the output
- the MC information for both D* and Lc

### CODE REVIEW

File: PWGHF/D2H/Tasks/taskCharmPolarisation.cxx [hunk 1]
```diff
@@ -217,20 +326,210 @@ struct TaskPolarisationCharmHadrons {
...
+  /// \param invMassCharmHad is the invariant-mass of the candidate
+  /// \param ptCharmHad is the pt of the candidate
+  /// \param pzCharmHad is the pz of the candidate
+  /// \param rapCharmHad is the rapidity of the candidate
+  /// \param cosThetaStar is the cosThetaStar of the candidate
+  /// \param outputMl is the array with ML output scores
+  /// \param isRotatedCandidate is a flag that keeps the info of the rotation of the candidate for bkg studies
+  /// \param origin is the MC origin
+  template <charm_polarisation::DecayChannel channel, bool withMl, bool doMc, charm_polarisation::CosThetaStarType cosThetaStarType>
+  void fillRecoHistos(float& invMassCharmHad, float& ptCharmHad, float& pzCharmHad, float& rapCharmHad, float& cosThetaStar, std::array<float, 3>& outputMl, int& isRotatedCandidate, int8_t origin)
```
Comment (mfaggin) [inline PWGHF/D2H/Tasks/taskCharmPolarisation.cxx:338]:
```suggestion
  void fillRecoHistos(float invMassCharmHad, float ptCharmHad, float pzCharmHad, float rapCharmHad, float cosThetaStar, std::array<float, 3> outputMl, int isRotatedCandidate, int8_t origin)
```

similar comment done by Vit in this PR https://github.com/AliceO2Group/O2Physics/pull/5057 ("Passing small types by value is better")

File: PWGHF/D2H/Tasks/taskCharmPolarisation.cxx [hunk 2]
```diff
@@ -217,20 +326,210 @@ struct TaskPolarisationCharmHadrons {
...
+    }
+  }
+
+  /// \param ptCharmHad is the pt of the particle
+  /// \param pzCharmHad is the pz of the particle
+  /// \param rapCharmHad is the rapidity of the particle
+  /// \param cosThetaStar is the cosThetaStar of the particle
+  /// \param origin is the MC origin
+  template <charm_polarisation::CosThetaStarType cosThetaStarType>
+  void fillGenHistos(float& ptCharmHad, float& pzCharmHad, float& rapCharmHad, float& cosThetaStar, int8_t origin)
```
Comment (mfaggin) [inline PWGHF/D2H/Tasks/taskCharmPolarisation.cxx:469]:
```suggestion
  void fillGenHistos(float ptCharmHad, float pzCharmHad, float rapCharmHad, float cosThetaStar, int8_t origin)
```

same as above

File: PWGHF/D2H/Tasks/taskCharmPolarisation.cxx [hunk 3]
```diff
@@ -217,20 +326,210 @@ struct TaskPolarisationCharmHadrons {
...
+  {
+    if constexpr (cosThetaStarType == charm_polarisation::CosThetaStarType::Helicity) { // Helicity
+      if (origin == RecoDecay::OriginType::Prompt) {                                    // prompt
+        registry.fill(HIST("hGenPromptHelicity"), ptCharmHad, pzCharmHad, rapCharmHad, cosThetaStar);
+      } else { // non-prompt
+        registry.fill(HIST("hGenNonPromptHelicity"), ptCharmHad, pzCharmHad, rapCharmHad, cosThetaStar);
+      }
+    } else if constexpr (cosThetaStarType == charm_polarisation::CosThetaStarType::Production) { // Prodution
```
Comment (mfaggin) [inline PWGHF/D2H/Tasks/taskCharmPolarisation.cxx:477]:
```suggestion
    } else if constexpr (cosThetaStarType == charm_polarisation::CosThetaStarType::Production) { // Production
```

File: PWGHF/D2H/Tasks/taskCharmPolarisation.cxx [hunk 4]
```diff
@@ -244,18 +543,30 @@ struct TaskPolarisationCharmHadrons {
         pyCharmHad = candidate.pyDstar();
         pzCharmHad = candidate.pzDstar();
         massDau = massPi; // (*)
-        invMassCharmHad = (candidate.signSoftPi() > 0) ? candidate.invMassDstar() : candidate.invMassAntiDstar();
-        invMassCharmHadForSparse = (candidate.signSoftPi() > 0) ? (invMassCharmHad - candidate.invMassD0()) : (invMassCharmHad - candidate.invMassD0Bar()); // different for D*
+        auto sign = candidate.signSoftPi();
+        if (sign > 0) {
+          invMassCharmHad = candidate.invMassDstar();
```
Comment (mfaggin) [inline PWGHF/D2H/Tasks/taskCharmPolarisation.cxx:548]:
```suggestion
        if (candidate.signSoftPi() > 0) {
          invMassCharmHad = candidate.invMassDstar();
```

**Review by mfaggin — CHANGES_REQUESTED** [review]
thank you very much @fgrosa , a few minor ocmments from my side

**Comment by fgrosa** [comment]
Hi @mfaggin thanks for the comments, all is implemented!

**Review by mfaggin — APPROVED** [review]

**Pull request merged**